pkg/x/importer/generate.go

```diff
@@ -0,0 +1,170 @@
+package importer
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2/model"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2/syntax"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"
+	"github.com/zclconf/go-cty/cty"
+)
+
+type LanguageGenerator func(w io.Writer, p *hcl2.Program) error
+
+func GenerateLanguageDefinition(w io.Writer, host plugin.Host, gen LanguageGenerator, state *resource.State) error {
+	hcl2Def, err := GenerateHCL2Definition(state)
+	if err != nil {
+		return err
+	}
+
+	text := fmt.Sprintf("%v", hcl2Def)
+
+	parser := syntax.NewParser()
+	if err = parser.ParseFile(strings.NewReader(text), string(state.URN.Name()+".pp")); err != nil {
+		return err
+	}
+	contract.Assert(len(parser.Diagnostics) == 0)
+
+	program, diags, err := hcl2.BindProgram(parser.Files, hcl2.PluginHost(host))
+	if err != nil {
+		return err
+	}
+	if len(diags) != 0 {
+		// TODO: do better here
+		return diags[0]
```

Comment:
This should be treated as an internal error, and ideally should capture the diagnostics.